esp-wifi/src/ble/npl.rs

```diff
@@ -880,36 +853,35 @@ unsafe extern "C" fn ble_npl_eventq_remove(
     queue: *const ble_npl_eventq,
     event: *const ble_npl_event,
 ) {
-    trace!("ble_npl_eventq_remove {:?} {:?}", queue, event);
+    info!("ble_npl_eventq_remove {:?} {:?}", queue, event);
 
-    if (*queue).dummy == 0 {
-        panic!("Try to use uninitialized queue");
-    }
+    core::assert!((*queue).dummy != 0);
+    let evt = (*event).dummy as *mut Event;
+    core::assert!(!evt.is_null());
 
-    if (*event).dummy == 0 {
-        panic!("Try to use uninitialized event");
+    if !(*evt).queued {
+        return;
     }
 
-    critical_section::with(|_| {
-        unwrap!(EVENTS[((*event).dummy - 1) as usize].as_mut()).queued = false;
-    });
+    let queue = (*queue).dummy as *mut ConcurrentQueue;
+    (*queue).remove(addr_of!(event) as *mut _);
+
+    (*evt).queued = false;
 }
 
 unsafe extern "C" fn ble_npl_eventq_put(queue: *const ble_npl_eventq, event: *const ble_npl_event) {
     trace!("ble_npl_eventq_put {:?} {:?}", queue, event);
 
-    if (*queue).dummy == 0 {
-        panic!("Try to use uninitialized queue");
-    }
+    core::assert!((*queue).dummy != 0);
 
-    if (*event).dummy == 0 {
-        panic!("Try to use uninitialized event");
-    }
+    let evt = (*event).dummy as *mut Event;
+    core::assert!(!evt.is_null());
 
-    critical_section::with(|_| {
-        unwrap!(EVENTS[((*event).dummy - 1) as usize].as_mut()).queued = true;
-        unwrap!(EVENT_QUEUE.enqueue((*event).dummy as usize));
-    });
+    (*evt).queued = true;
+
+    let queue = (*queue).dummy as *mut ConcurrentQueue;
+    let mut event = event as usize;
+    (*queue).enqueue(addr_of_mut!(event).cast());
```

Comment:
This looks wrong, this passes the address of the parameter (i.e. points to the stack), not the event pointer itself. Same for `remove`, unless I'm missing something.

Reply:
Ah I see

`enqueue` and `remove` take a pointer to some data of length `item_size` and copy from that location into the queue.
The NPL stuff however wants us to enqueue and dequeue a pointer

(Hope that explanation makes sense - not sure how to phrase it differently)

Reply:
So what happens, we write the address of the local `event` to the queue, and return. The return invalidates the pointer that we just queued. What happens next and why don't we crash?

Reply:
Yes - RawQueue was initially done to satisfy what the WiFi driver expects from a queue - in an upcoming PR we should use something other than RawQueue here for NPL (and looking back I should have split things into separate more granular PRs)